src/Riok.Mapperly/Descriptors/Enumerables/EnsureCapacity/EnsureCapacityBuilder.cs

```diff
@@ -0,0 +1,69 @@
+﻿using System.Diagnostics.CodeAnalysis;
+using Microsoft.CodeAnalysis;
+using Riok.Mapperly.Helpers;
+
+namespace Riok.Mapperly.Descriptors.Enumerables.EnsureCapacity;
+
+public static class EnsureCapacityBuilder
+{
+    private const string EnsureCapacityName = "EnsureCapacity";
+    private const string CountPropertyName = nameof(ICollection<object>.Count);
+    private const string LengthPropertyName = nameof(Array.Length);
+    private const string TryGetNonEnumeratedCountMethodName = "TryGetNonEnumeratedCount";
+
+    public static EnsureCapacity? TryBuildEnsureCapacity(ITypeSymbol sourceType, ITypeSymbol targetType, WellKnownTypes types)
+    {
+        var capacityMethod = targetType.GetAllMembers(EnsureCapacityName)
+                        .OfType<IMethodSymbol>()
+                        .FirstOrDefault(x => x.Parameters.Length == 1
+                        && x.Parameters[0].Type.SpecialType == SpecialType.System_Int32
+                        && !x.IsStatic);
+
+        // if EnsureCapacity is not available then return null
+        if (capacityMethod == null)
+            return null;
+
+        // if target does not have a count then return null
+        if (!TryGetNonEnumeratedCount(targetType, types, out var targetSizeProperty))
+            return null;
+
+        // if target and source count are known then create a simple EnsureCapacity statement
+        if (TryGetNonEnumeratedCount(sourceType, types, out var sourceSizeProperty))
+            return new EnsureCapacityMember(targetSizeProperty, sourceSizeProperty);
+
+        if (!sourceType.ImplementsGeneric(types.IEnumerableT, out var iEnumerable))
+            return null;
+
+        var nonEnumeratedCountMethod = types.Enumerable.GetMembers(TryGetNonEnumeratedCountMethodName)
+                                                        .OfType<IMethodSymbol>()
+                                                        .FirstOrDefault(x => x.ReturnType.SpecialType == SpecialType.System_Boolean && x.IsStatic && x.Parameters.Length == 2 && x.IsGenericMethod);
+
+        // if source does not have a count use GetNonEnumeratedCount, calling EnusureCapacity if count is available
+        var typedNonEnumeratedCount = nonEnumeratedCountMethod.Construct(iEnumerable.TypeArguments.ToArray());
```

Comment:
This leads to a null ref exception on .NET Framework... add a null guard:
```c#
if (nonEnumeratedCountMethod == null)
    return null;

...
```

Reply:
.NET Framework can't get to this section, [line 23](https://github.com/riok/mapperly/pull/312/files/f19efb7c50ed837ab5986cf93b4ac58f64072d96#diff-3f8932b41774a45b65c962c4ae183d71993fe8112d68981d3ab20f2a1d53172fR23) prevents non .NET 6+ code from getting here because they don't have `EnsureCapacity` available yet.